### PR TITLE
Module aws_organizations_facts: added module

### PR DESCRIPTION
**Pull request by seansummers**
##### SUMMARY
Added `aws_organizations_facts` module.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

Obtain facts about AWS Organizations

##### ADDITIONAL INFORMATION

Gets information about AWS Organizations

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/aws_organizations_facts.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+def main():
+    argument_spec = dict(all_facts=dict(type='bool',
+                                        required=False,
+                                        default=False),
+                         accounts=dict(type='bool',
+                                       required=False,
+                                       default=False))
+    module = AnsibleAWSModule(argument_spec=argument_spec,
+                              supports_check_mode=True)
+    if not HAS_BOTO3:
```
Comment (flowerysong) [inline lib/ansible/modules/cloud/amazon/aws_organizations_facts.py]: This is unnecessary; AnsibleAWSModule will already have terminated module execution if boto3 is not available.

**Review by flowerysong — CHANGES_REQUESTED** [review]
I don't think this should be a facts module, since it's gathering information tied to a set of credentials rather than the machine it's running on (or any specific host). I suggest renaming it to `aws_organization_info`, and returning the info as a top-level `organization` dict instead of under `ansible_facts`.  (This will only return a single organization, so the pluralisation should be avoided.)

Integration tests would also be nice.

File: lib/ansible/modules/cloud/amazon/aws_organizations_facts.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    result = {'ansible_facts': {}}
+
+    try:
+        organizations = describe_organization(client)
+        result['ansible_facts']['organizations'] = camel_dict_to_snake_dict(
+            organizations)
+
+        # Optional outputs
+        all_facts = module.params.get('all_facts')
+        if all_facts or module.params.get('accounts'):
```
Comment (flowerysong) [inline lib/ansible/modules/cloud/amazon/aws_organizations_facts.py:134]: Why do you have two separate parameters with identical functionality?

File: lib/ansible/modules/cloud/amazon/aws_organizations_facts.py [hunk 3]
```diff
@@ -93,12 +93,10 @@
           name: Master AWS Account
           status: ACTIVE
 '''
-from ansible.module_utils.aws.core import (AnsibleAWSModule, HAS_BOTO3)
+from ansible.module_utils.aws.core import AnsibleAWSModule
 from ansible.module_utils.ec2 import (AWSRetry, camel_dict_to_snake_dict)
-try:
```
Comment (flowerysong) [inline lib/ansible/modules/cloud/amazon/aws_organizations_facts.py]: Imports run before AnsibleAWSModule instantiation, so you still need this try.

**Comment by seansummers** [comment]
@flowerysong 
> I don't think this should be a facts module, since it's gathering information tied to a set of credentials rather than the machine it's running on (or any specific host). I suggest renaming it to `aws_organization_info`, and returning the info as a top-level `organization` dict instead of under `ansible_facts`. (This will only return a single organization, so the pluralisation should be avoided.)

I'm uncomfortable with creating the first aws_*_info module in the mainline. I also feel you are describing this module as a duplicate of `aws_caller_facts`, which it is not.

The facts this module returns provides a basis for life-cycling accounts in an AWS organization, including:
- Organizational Units
- Creating, Inviting, Accepting, and Removing Accounts from an Organization
- Managing account membership in OUs
- Managing SCPs and applying them to OUs and Accounts

If coverage of this functionality exists in Ansible today, please point me to the reference, and I'll remove this PR.

**Comment by flowerysong** [comment]
See #57274; most AWS _facts modules have been renamed to _info in devel already.

**Comment by seansummers** [comment]
Thank you, @flowerysong , and I believe I understand better now.

The purpose of the `aws_organizations_*` modules I was working on is to manage AWS Organization Accounts and the SCPs that apply to them, which would probably require treating AWS Accounts as first class manageable "host machines" rather than an attribute of a machine.

There does not seem to be much desire/need in the Ansible community to support this multi-account AWS Organizations and service orchestration on the cloud controle plane, above the level of host-based IPv4/6 connectivity. AWS has released AWS Control Tower as a non-api SAAS to provide this functionality where current tooling does not exist in the open source community.

I now feel that this module is better suited as an inventory plugin, with OUs as "groups" and accounts as "host machines", if development is to be continued.

**Comment by flowerysong** [comment]
It seems to fit perfectly well as an info module, to me. I don't see any reason turning it into an inventory plugin would be useful.

**Comment by felixfontein** [comment]
This definitely must be an `_info` module and must not return `ansible_facts`. The remaining `_facts` modules for cloud services will be renamed eventually, this mainly takes more time becaues they actually use `ansible_facts` and deprecating return values isn't easily done (yet).

**Comment by felixfontein** [comment]
(I can't really comment on whether this being inventory plugin would make more sense, since I'm not familiar with AWS OUs; but if it should be one of `_info` or `_facts`, it should definitely be `_info`.)

File: lib/ansible/modules/cloud/amazon/aws_organizations_facts.py (hunk 2 above)
Comment (tremble) [reply lib/ansible/modules/cloud/amazon/aws_organizations_facts.py:134]: This looks like the logic from one of the other _info/_fact modules (I forget which).  The idea is that in the long run you add
- accounts
- policies
- (something else)

and all_facts lists them all.  The advantage here comes when you don't have full read-access, and can only read certain attributes.  I'm not sure it makes sense without also supporting listing the SCPs attached to the OU

**Comment by tremble** [comment]
@felixfontein almost certainly an _info module.  I can't see how this would fit into an inventory plugin...

**Pull request closed without merging**